src/processes.c

```diff
@@ -46,76 +46,220 @@
+ * can be terminated
+ *
+ * @param pid id of the child process
+ * @return int 0 on success, NOCHILD if child does not exist, NOINIT if not
+ * initialized
  */
 int
 terminate_process (pid_t pid)
 {
-  int ret;
-
-  if (pid == 0)
-    return 0;
+  if (initialized)
+    {
+      for (int i = 0; i < max_procs; i++)
+        {
+          if (procs[i].pid == pid)
+            {
+              kill (pid, SIGTERM);
+              usleep (10000);
+              if (!procs[i].terminated)
+                kill (pid, SIGKILL);
+              return 0;
+            }
+        }
+      return NOCHILD;
+    }
+  else
+    {
+      kill (pid, SIGTERM);
+      usleep (10000);
+      if (waitpid (pid, NULL, WNOHANG))
+        kill (pid, SIGKILL);
+      return 0;
+    }
+}
 
-  ret = kill (pid, SIGTERM);
+/**
+ * @brief This function terminates all processes spawned with create_process.
+ * Calls terminate_child for each process active. In case init_procs was not
+ * called this function does nothing.
+ *
+ */
+void
+procs_terminate_childs ()
+{
+  if (!initialized)
+    return;
 
-  if (ret == 0)
+  for (int i = 0; i < max_procs; i++)
     {
-      usleep (1000);
+      if (!procs[i].terminated)
+        terminate_process (procs[i].pid);
+    }
+}
 
-      if (waitpid (pid, NULL, WNOHANG) >= 0)
-        {
-          kill (pid, SIGKILL);
-          return -1;
-        }
+/**
+ * @brief Handler for a termination signal. This will terminate all childs and
+ * calls SIGTERM for itself afterwards.
+ *
+ */
+static void
+terminate ()
+{
+  if (!initialized)
+    return;
+
+  int tries = 5;
+  procs_terminate_childs ();
+
+  while (num_procs && tries--)
+    {
+      sleep (1);
     }
 
-  return 0;
+  clean_procs ();
+  openvas_signal (SIGTERM, SIG_DFL);
+  raise (SIGTERM);
+}
+
+/**
+ * @brief Init procs, must be called once per process
+ *
+ * @param max
+ */
+void
+procs_init (int max)
+{
+  procs = g_malloc (max * sizeof (proc_entry));
+  for (int i = 0; i < max; i++)
+    {
+      procs[i].terminated = 1;
+    }
+  max_procs = max;
+  num_procs = 0;
+  openvas_signal (SIGCHLD, clear_child);
+  openvas_signal (SIGTERM, terminate);
+  initialized = 1;
 }
 
 static void
 init_child_signal_handlers (void)
 {
   /* SIGHUP is only for reloading main scanner process. */
   openvas_signal (SIGHUP, SIG_IGN);
-  openvas_signal (SIGTERM, make_em_die);
-  openvas_signal (SIGINT, make_em_die);
-  openvas_signal (SIGQUIT, make_em_die);
   openvas_signal (SIGSEGV, sighand_segv);
   openvas_signal (SIGPIPE, SIG_IGN);
 }
 
 /**
- * @brief Create a new process (fork).
+ * @brief Calls a function with a new process
+ *
+ * @param func Function to call
+ * @param args arguments
+ * @return pid of spawned process on success or one of the following errors:
+ * FORKFAILED, NOINIT, PROCSFULL
  */
 pid_t
-create_process (process_func_t function, void *argument)
+create_process (process_func_t func, void *args)
 {
-  int pid;
-
-  gvm_log_lock ();
```

Comment:
Why did you remove this? This code was introduced some time ago to avoid deadlock. See PR #491